<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DO Dot plot — DO.Dotplot • DOtools</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DO Dot plot — DO.Dotplot"><meta name="description" content="This function generates a dot plot for multiple genes, comparing expression levels across one or two specified groups.
It supports both individual and pseudobulk expression calculations.
Highly variable customization options allow control over dot size, color scaling, annotations, and axis orientation.
The function integrates seamlessly with Seurat objects for single-cell RNA-seq analysis."><meta property="og:description" content="This function generates a dot plot for multiple genes, comparing expression levels across one or two specified groups.
It supports both individual and pseudobulk expression calculations.
Highly variable customization options allow control over dot size, color scaling, annotations, and axis orientation.
The function integrates seamlessly with Seurat objects for single-cell RNA-seq analysis."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DO Dot plot</h1>

      <div class="d-none name"><code>DO.Dotplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function generates a dot plot for multiple genes, comparing expression levels across one or two specified groups.
It supports both individual and pseudobulk expression calculations.
Highly variable customization options allow control over dot size, color scaling, annotations, and axis orientation.
The function integrates seamlessly with Seurat objects for single-cell RNA-seq analysis.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DO.Dotplot</span><span class="op">(</span></span>
<span>  <span class="va">Seu_object</span>,</span>
<span>  <span class="va">Feature</span>,</span>
<span>  group.by.x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group.by.y <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group.by.y2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  across.group.by.x <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  across.group.by.y <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  sort_x <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dot.size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>  plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  midpoint <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  scale_gene <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  returnValue <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  log1p_nUMI <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  hide_zero <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  annotation_x <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  annotation_x_position <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  annotation_x_rev <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  point_stroke <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>  limits_colorscale <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  coord_flip <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seu-object">Seu_object<a class="anchor" aria-label="anchor" href="#arg-seu-object"></a></dt>
<dd><p>The seurat Seu_object</p></dd>


<dt id="arg-feature">Feature<a class="anchor" aria-label="anchor" href="#arg-feature"></a></dt>
<dd><p>Genes or DF of interest, Data frame should have columns with gene and annotation information, e.g. output of FindAllMarkers</p></dd>


<dt id="arg-group-by-x">group.by.x<a class="anchor" aria-label="anchor" href="#arg-group-by-x"></a></dt>
<dd><p>group name to plot on x-axis</p></dd>


<dt id="arg-group-by-y">group.by.y<a class="anchor" aria-label="anchor" href="#arg-group-by-y"></a></dt>
<dd><p>group name to look for in meta data</p></dd>


<dt id="arg-group-by-y-">group.by.y2<a class="anchor" aria-label="anchor" href="#arg-group-by-y-"></a></dt>
<dd><p>second group name to look for in meta data</p></dd>


<dt id="arg-across-group-by-x">across.group.by.x<a class="anchor" aria-label="anchor" href="#arg-across-group-by-x"></a></dt>
<dd><p>calculate a pseudobulk expression approach for the x-axis categories</p></dd>


<dt id="arg-across-group-by-y">across.group.by.y<a class="anchor" aria-label="anchor" href="#arg-across-group-by-y"></a></dt>
<dd><p>calculate a pseudobulk expression approach for the y-axis categories, if group.by.y2 is specified Pseudobulk will be splitted by it</p></dd>


<dt id="arg-sort-x">sort_x<a class="anchor" aria-label="anchor" href="#arg-sort-x"></a></dt>
<dd><p>Vector sorting the xaxis</p></dd>


<dt id="arg-dot-size">dot.size<a class="anchor" aria-label="anchor" href="#arg-dot-size"></a></dt>
<dd><p>Vector of dot size</p></dd>


<dt id="arg-plot-margin">plot.margin<a class="anchor" aria-label="anchor" href="#arg-plot-margin"></a></dt>
<dd><p>= plot margins</p></dd>


<dt id="arg-midpoint">midpoint<a class="anchor" aria-label="anchor" href="#arg-midpoint"></a></dt>
<dd><p>midpoint in color gradient</p></dd>


<dt id="arg-scale-gene">scale_gene<a class="anchor" aria-label="anchor" href="#arg-scale-gene"></a></dt>
<dd><p>IF True calculates the Z-score of the average expression per gene</p></dd>


<dt id="arg-hide-zero">hide_zero<a class="anchor" aria-label="anchor" href="#arg-hide-zero"></a></dt>
<dd><p>Removes dots for genes with 0 expression</p></dd>


<dt id="arg-annotation-x">annotation_x<a class="anchor" aria-label="anchor" href="#arg-annotation-x"></a></dt>
<dd><p>Adds annotation on top of x axis instead on y axis</p></dd>


<dt id="arg-point-stroke">point_stroke<a class="anchor" aria-label="anchor" href="#arg-point-stroke"></a></dt>
<dd><p>Defines the thickness of the black stroke on the dots</p></dd>


<dt id="arg-limits-colorscale">limits_colorscale<a class="anchor" aria-label="anchor" href="#arg-limits-colorscale"></a></dt>
<dd><p>Set manually colorscale limits</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Further arguments passed to annoSegment function if annotation_x == T</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a ggplot</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mariano Ruz Jurado</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">DO.Dotplot</span><span class="op">(</span></span></span>
<span class="r-in"><span>  Seu_object <span class="op">=</span> <span class="va">Seurat</span>,</span></span>
<span class="r-in"><span>  Feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDH5"</span>,<span class="st">"TTN"</span>,<span class="st">"MALAT1"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  group.by.x<span class="op">=</span><span class="st">"condition"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

