<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DO.enrichR — DO.enrichR • DOtools</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DO.enrichR — DO.enrichR"><meta name="description" content="Performs Gene Ontology enrichment analysis on differentially expressed genes using the EnrichR API.
Separately analyzes upregulated and downregulated genes and returns results."><meta property="og:description" content="Performs Gene Ontology enrichment analysis on differentially expressed genes using the EnrichR API.
Separately analyzes upregulated and downregulated genes and returns results."><script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adfct.html">Usage of functions</a></li>
    <li><a class="dropdown-item" href="../articles/cb.html">Remove ambient RNA with CellBender</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DO.enrichR</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarianoRuzJurado/DOtools/blob/master/R/utils_functions.R" class="external-link"><code>R/utils_functions.R</code></a></small>
      <div class="d-none name"><code>DO.enrichR.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs Gene Ontology enrichment analysis on differentially expressed genes using the EnrichR API.
Separately analyzes upregulated and downregulated genes and returns results.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DO.enrichR</span><span class="op">(</span></span>
<span>  <span class="va">df_DGE</span>,</span>
<span>  <span class="va">gene_column</span>,</span>
<span>  <span class="va">pval_column</span>,</span>
<span>  <span class="va">log2fc_column</span>,</span>
<span>  pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  log2fc_cutoff <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">""</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Human"</span>,</span>
<span>  go_catgs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO_Molecular_Function_2023"</span>, <span class="st">"GO_Cellular_Component_2023"</span>,</span>
<span>    <span class="st">"GO_Biological_Process_2023"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-df-dge">df_DGE<a class="anchor" aria-label="anchor" href="#arg-df-dge"></a></dt>
<dd><p>data.frame containing differential gene expression results.</p></dd>


<dt id="arg-gene-column">gene_column<a class="anchor" aria-label="anchor" href="#arg-gene-column"></a></dt>
<dd><p>column name in <code>df</code> with gene symbols.</p></dd>


<dt id="arg-pval-column">pval_column<a class="anchor" aria-label="anchor" href="#arg-pval-column"></a></dt>
<dd><p>column name in <code>df</code> with p-values.</p></dd>


<dt id="arg-log-fc-column">log2fc_column<a class="anchor" aria-label="anchor" href="#arg-log-fc-column"></a></dt>
<dd><p>column name in <code>df</code> with log2 fold changes.</p></dd>


<dt id="arg-pval-cutoff">pval_cutoff<a class="anchor" aria-label="anchor" href="#arg-pval-cutoff"></a></dt>
<dd><p>adjusted p-value threshold for significance (default = 0.05).</p></dd>


<dt id="arg-log-fc-cutoff">log2fc_cutoff<a class="anchor" aria-label="anchor" href="#arg-log-fc-cutoff"></a></dt>
<dd><p>log2 fold change threshold for up/down regulation (default = 0.25).</p></dd>


<dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>folder path where the output Excel file will be saved. A subfolder "GSA_Tables" will be created.</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>suffix used in the Excel filename (e.g., "GSA_CellType_MyAnalysis.xlsx").</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>species name for enrichment analysis. Options include "Human", "Mouse", "Yeast", etc. (default = "Mouse").</p></dd>


<dt id="arg-go-catgs">go_catgs<a class="anchor" aria-label="anchor" href="#arg-go-catgs"></a></dt>
<dd><p>GO databases to use. Defaults to c("GO_Molecular_Function_2023", "GO_Cellular_Component_2023", "GO_Biological_Process_2023").</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.frame with GO enrichment results if <code>path</code> is NULL, otherwise writes an Excel file.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">enrichR</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sce_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce_data.rds"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DGE_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="DO.MultiDGE.html">DO.MultiDGE</a></span><span class="op">(</span><span class="va">sce_data</span>,</span></span>
<span class="r-in"><span>                         sample_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span></span>
<span class="r-in"><span>                         method_sc <span class="op">=</span> <span class="st">"wilcox"</span>,</span></span>
<span class="r-in"><span>                         annotation_col <span class="op">=</span> <span class="st">"annotation"</span>,</span></span>
<span class="r-in"><span>                         ident_ctrl <span class="op">=</span> <span class="st">"healthy"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Centering and scaling data matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:17 - Corrected annotation names in pseudo-bulk object by replacing '-' with '_'.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:17 - Starting DGE single cell method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:17 - Comparing disease with healthy in: Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:17 - Skipping Monocytes since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:17 - Comparing disease with healthy in: CD4_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:20 - Comparing disease with healthy in: NK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:21 - Comparing disease with healthy in: CD8_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:21 - Skipping CD8_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:21 - Comparing disease with healthy in: B_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:21 - Comparing disease with healthy in: pDC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Finished DGE single cell method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Starting DGE pseudo bulk method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: B_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping B_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: CD4_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping CD4_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping Monocytes since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: NK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping NK since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: pDC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping pDC since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Comparing disease with healthy in: CD8_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Skipping CD8_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - Finished DGE pseudo bulk method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-08-06 09:47:22 - DGE pseudo bulk result is empty...</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">DGE_result</span> <span class="op">&lt;-</span> <span class="va">DGE_result</span><span class="op">[</span><span class="va">DGE_result</span><span class="op">$</span><span class="va">celltype</span> <span class="op">==</span> <span class="st">"CD4_T_cells"</span>,<span class="op">]</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">result_GO</span> <span class="op">&lt;-</span> <span class="fu">DO.enrichR</span><span class="op">(</span>df_DGE <span class="op">=</span> <span class="va">DGE_result</span>,</span></span>
<span class="r-in"><span>                       gene_column <span class="op">=</span> <span class="st">"gene"</span>,</span></span>
<span class="r-in"><span>                       pval_column <span class="op">=</span> <span class="st">"p_val_SC_wilcox"</span>,</span></span>
<span class="r-in"><span>                       log2fc_column <span class="op">=</span> <span class="st">"avg_log2FC_SC_wilcox"</span>,</span></span>
<span class="r-in"><span>                       pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span></span>
<span class="r-in"><span>                       log2fc_cutoff <span class="op">=</span> <span class="fl">0.25</span>,</span></span>
<span class="r-in"><span>                       path <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>                       filename <span class="op">=</span> <span class="st">""</span>,</span></span>
<span class="r-in"><span>                       species <span class="op">=</span> <span class="st">"Human"</span>,</span></span>
<span class="r-in"><span>                       go_catgs <span class="op">=</span> <span class="st">"GO_Biological_Process_2023"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Connection changed to https://maayanlab.cloud/Enrichr/</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Connection is Live!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Uploading data to Enrichr... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Querying GO_Biological_Process_2023... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parsing results... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Uploading data to Enrichr... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   Querying GO_Biological_Process_2023... Done.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Parsing results... Done.</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

