<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DO Heatmap of the mean expression of genes across a groups — DO.Heatmap • DOtools</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DO Heatmap of the mean expression of genes across a groups — DO.Heatmap"><meta name="description" content="Wrapper around heatmap.py, which generates a heatmap of
showing the average nUMI for a set of genes in different groups.
Addiional an argumnt can be made to show foldchanges between two conditions.
Differential gene expression analysis between the different groups can be
performed."><meta property="og:description" content="Wrapper around heatmap.py, which generates a heatmap of
showing the average nUMI for a set of genes in different groups.
Addiional an argumnt can be made to show foldchanges between two conditions.
Differential gene expression analysis between the different groups can be
performed."><script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DO Heatmap of the mean expression of genes across a groups</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarianoRuzJurado/DOtools/blob/master/R/DO.Heatmap.R" class="external-link"><code>R/DO.Heatmap.R</code></a></small>
      <div class="d-none name"><code>DO.Heatmap.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Wrapper around heatmap.py, which generates a heatmap of
showing the average nUMI for a set of genes in different groups.
Addiional an argumnt can be made to show foldchanges between two conditions.
Differential gene expression analysis between the different groups can be
performed.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DO.Heatmap</span><span class="op">(</span></span>
<span>  <span class="va">sce_object</span>,</span>
<span>  <span class="va">features</span>,</span>
<span>  assay_normalized <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>  groups_order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  value_plot <span class="op">=</span> <span class="st">"expr"</span>,</span>
<span>  group_fc <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>  group_fc_ident_1 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_fc_ident_2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip_value <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max_fc <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  z_score <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  filename <span class="op">=</span> <span class="st">"Heatmap.svg"</span>,</span>
<span>  swap_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cmap <span class="op">=</span> <span class="st">"Reds"</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title_fontprop <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clustering_method <span class="op">=</span> <span class="st">"complete"</span>,</span>
<span>  clustering_metric <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  cluster_x_axis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_y_axis <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  axs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  figsize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span>,</span>
<span>  linewidth <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>  ticks_fontdict <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xticks_rotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  yticks_rotation <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vmin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  vcenter <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  vmax <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_title <span class="op">=</span> <span class="st">"LogMean(nUMI)\nin group"</span>,</span>
<span>  add_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  df_pvals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  stats_x_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  square_x_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  test <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  log2fc_cutoff <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  only_pos <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  square <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  showP <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  logcounts <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sce-object">sce_object<a class="anchor" aria-label="anchor" href="#arg-sce-object"></a></dt>
<dd><p>SCE object or Seurat with meta.data</p></dd>


<dt id="arg-features">features<a class="anchor" aria-label="anchor" href="#arg-features"></a></dt>
<dd><p>gene names or continuous value in meta data</p></dd>


<dt id="arg-assay-normalized">assay_normalized<a class="anchor" aria-label="anchor" href="#arg-assay-normalized"></a></dt>
<dd><p>Assay with raw counts</p></dd>


<dt id="arg-group-by">group_by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>meta data column name with categorical values</p></dd>


<dt id="arg-groups-order">groups_order<a class="anchor" aria-label="anchor" href="#arg-groups-order"></a></dt>
<dd><p>order for the categories in the group_by</p></dd>


<dt id="arg-value-plot">value_plot<a class="anchor" aria-label="anchor" href="#arg-value-plot"></a></dt>
<dd><p>plotted values correspond to expression values or
foldchanges</p></dd>


<dt id="arg-group-fc">group_fc<a class="anchor" aria-label="anchor" href="#arg-group-fc"></a></dt>
<dd><p>if foldchanges specified than the groups must be specified
that will be compared</p></dd>


<dt id="arg-group-fc-ident-">group_fc_ident_1<a class="anchor" aria-label="anchor" href="#arg-group-fc-ident-"></a></dt>
<dd><p>Defines the first group in the test</p></dd>


<dt id="arg-group-fc-ident-">group_fc_ident_2<a class="anchor" aria-label="anchor" href="#arg-group-fc-ident-"></a></dt>
<dd><p>Defines the second group in the test</p></dd>


<dt id="arg-clip-value">clip_value<a class="anchor" aria-label="anchor" href="#arg-clip-value"></a></dt>
<dd><p>Clips the colourscale to the 99th percentile, useful if
one gene is driving the colourscale</p></dd>


<dt id="arg-max-fc">max_fc<a class="anchor" aria-label="anchor" href="#arg-max-fc"></a></dt>
<dd><p>Clips super high foldchanges to this value, so changes can
still be appreciated</p></dd>


<dt id="arg-z-score">z_score<a class="anchor" aria-label="anchor" href="#arg-z-score"></a></dt>
<dd><p>apply z-score transformation, "group" or "var"</p></dd>


<dt id="arg-path">path<a class="anchor" aria-label="anchor" href="#arg-path"></a></dt>
<dd><p>path to save the plot</p></dd>


<dt id="arg-filename">filename<a class="anchor" aria-label="anchor" href="#arg-filename"></a></dt>
<dd><p>name of the file</p></dd>


<dt id="arg-swap-axes">swap_axes<a class="anchor" aria-label="anchor" href="#arg-swap-axes"></a></dt>
<dd><p>whether to swap the axes or not</p></dd>


<dt id="arg-cmap">cmap<a class="anchor" aria-label="anchor" href="#arg-cmap"></a></dt>
<dd><p>color map</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>title for the main plot</p></dd>


<dt id="arg-title-fontprop">title_fontprop<a class="anchor" aria-label="anchor" href="#arg-title-fontprop"></a></dt>
<dd><p>font properties for the title
(e.g., 'weight' and 'size')</p></dd>


<dt id="arg-clustering-method">clustering_method<a class="anchor" aria-label="anchor" href="#arg-clustering-method"></a></dt>
<dd><p>clustering method to use when hierarchically
clustering the x and y-axis</p></dd>


<dt id="arg-clustering-metric">clustering_metric<a class="anchor" aria-label="anchor" href="#arg-clustering-metric"></a></dt>
<dd><p>metric to use when hierarchically clustering the x-
and y-axis</p></dd>


<dt id="arg-cluster-x-axis">cluster_x_axis<a class="anchor" aria-label="anchor" href="#arg-cluster-x-axis"></a></dt>
<dd><p>hierarchically clustering the x-axis</p></dd>


<dt id="arg-cluster-y-axis">cluster_y_axis<a class="anchor" aria-label="anchor" href="#arg-cluster-y-axis"></a></dt>
<dd><p>hierarchically clustering the y-axis</p></dd>


<dt id="arg-axs">axs<a class="anchor" aria-label="anchor" href="#arg-axs"></a></dt>
<dd><p>matplotlib axis</p></dd>


<dt id="arg-figsize">figsize<a class="anchor" aria-label="anchor" href="#arg-figsize"></a></dt>
<dd><p>figure size</p></dd>


<dt id="arg-linewidth">linewidth<a class="anchor" aria-label="anchor" href="#arg-linewidth"></a></dt>
<dd><p>line width for the border of cells</p></dd>


<dt id="arg-ticks-fontdict">ticks_fontdict<a class="anchor" aria-label="anchor" href="#arg-ticks-fontdict"></a></dt>
<dd><p>font properties for the x and y ticks
(e.g.,  'weight' and 'size')</p></dd>


<dt id="arg-xticks-rotation">xticks_rotation<a class="anchor" aria-label="anchor" href="#arg-xticks-rotation"></a></dt>
<dd><p>rotation of the x-ticks</p></dd>


<dt id="arg-yticks-rotation">yticks_rotation<a class="anchor" aria-label="anchor" href="#arg-yticks-rotation"></a></dt>
<dd><p>rotations of the y-ticks</p></dd>


<dt id="arg-vmin">vmin<a class="anchor" aria-label="anchor" href="#arg-vmin"></a></dt>
<dd><p>minimum value</p></dd>


<dt id="arg-vcenter">vcenter<a class="anchor" aria-label="anchor" href="#arg-vcenter"></a></dt>
<dd><p>center value</p></dd>


<dt id="arg-vmax">vmax<a class="anchor" aria-label="anchor" href="#arg-vmax"></a></dt>
<dd><p>maximum value</p></dd>


<dt id="arg-legend-title">legend_title<a class="anchor" aria-label="anchor" href="#arg-legend-title"></a></dt>
<dd><p>title for the color bar</p></dd>


<dt id="arg-add-stats">add_stats<a class="anchor" aria-label="anchor" href="#arg-add-stats"></a></dt>
<dd><p>add statistical annotation, will add a square with an '*'
in the center if the expression is significantly different in a group with
respect to the others</p></dd>


<dt id="arg-df-pvals">df_pvals<a class="anchor" aria-label="anchor" href="#arg-df-pvals"></a></dt>
<dd><p>dataframe with the p-values, should be gene x group or group
x gene in case of swap_axes is False</p></dd>


<dt id="arg-stats-x-size">stats_x_size<a class="anchor" aria-label="anchor" href="#arg-stats-x-size"></a></dt>
<dd><p>size of the asterisk</p></dd>


<dt id="arg-square-x-size">square_x_size<a class="anchor" aria-label="anchor" href="#arg-square-x-size"></a></dt>
<dd><p>size and thickness of the square percentual, vector</p></dd>


<dt id="arg-test">test<a class="anchor" aria-label="anchor" href="#arg-test"></a></dt>
<dd><p>test to use for test for significance</p></dd>


<dt id="arg-pval-cutoff">pval_cutoff<a class="anchor" aria-label="anchor" href="#arg-pval-cutoff"></a></dt>
<dd><p>cutoff for the p-value</p></dd>


<dt id="arg-log-fc-cutoff">log2fc_cutoff<a class="anchor" aria-label="anchor" href="#arg-log-fc-cutoff"></a></dt>
<dd><p>minimum cutoff for the log2FC</p></dd>


<dt id="arg-only-pos">only_pos<a class="anchor" aria-label="anchor" href="#arg-only-pos"></a></dt>
<dd><p>if set to TRUE, only use positive genes in the condition</p></dd>


<dt id="arg-square">square<a class="anchor" aria-label="anchor" href="#arg-square"></a></dt>
<dd><p>whether to make the cell square or not</p></dd>


<dt id="arg-showp">showP<a class="anchor" aria-label="anchor" href="#arg-showp"></a></dt>
<dd><p>if set to false return a dictionary with the axis</p></dd>


<dt id="arg-logcounts">logcounts<a class="anchor" aria-label="anchor" href="#arg-logcounts"></a></dt>
<dd><p>whether the input is logcounts or not</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Depending on <code>showP</code>, returns the plot if set to <code>TRUE</code> or a
dictionary with the axes.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mariano Ruz Jurado &amp; David Rodriguez Morales</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#'</span></span></span>
<span class="r-in"><span><span class="va">sce_data</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce_data.rds"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">DO.Heatmap</span><span class="op">(</span></span></span>
<span class="r-in"><span>sce_object <span class="op">=</span> <span class="va">sce_data</span>,</span></span>
<span class="r-in"><span>assay_normalized <span class="op">=</span> <span class="st">"RNA"</span>,</span></span>
<span class="r-in"><span>group_by<span class="op">=</span><span class="st">"seurat_clusters"</span>,</span></span>
<span class="r-in"><span>features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce_data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>z_score <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>path <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>filename <span class="op">=</span> <span class="st">"Heatmap.svg"</span>,</span></span>
<span class="r-in"><span>swap_axes <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>cmap <span class="op">=</span> <span class="st">"Reds"</span>,</span></span>
<span class="r-in"><span>title <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>title_fontprop <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>clustering_method <span class="op">=</span> <span class="st">"complete"</span>,</span></span>
<span class="r-in"><span>clustering_metric <span class="op">=</span> <span class="st">"euclidean"</span>,</span></span>
<span class="r-in"><span>cluster_x_axis <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>cluster_y_axis <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>axs <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>figsize <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">6</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>linewidth <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>ticks_fontdict <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>xticks_rotation <span class="op">=</span> <span class="fl">45</span>,</span></span>
<span class="r-in"><span>yticks_rotation <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>vmin <span class="op">=</span> <span class="fl">0.0</span>,</span></span>
<span class="r-in"><span>vcenter <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>vmax <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>legend_title <span class="op">=</span> <span class="st">"LogMean(nUMI)\nin group"</span>,</span></span>
<span class="r-in"><span>add_stats <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>df_pvals <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>stats_x_size <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>square_x_size <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>test <span class="op">=</span> <span class="st">"wilcox"</span>,</span></span>
<span class="r-in"><span>pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span></span>
<span class="r-in"><span>log2fc_cutoff <span class="op">=</span> <span class="fl">0</span>,</span></span>
<span class="r-in"><span>only_pos <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>square <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>showP <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>logcounts <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 1</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No features pass logfc.threshold threshold; returning empty data.frame</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (default method for FindMarkers) please install the presto package</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> install.packages('devtools')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> devtools::install_github('immunogenomics/presto')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After installation of presto, Seurat will automatically use the more </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> efficient implementation (no further action necessary).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message will be shown once per session</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 3</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 4</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 5</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 6</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 7</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>No features pass logfc.threshold threshold; returning empty data.frame</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Calculating cluster 8</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John, DFG SFB 1366, Project B04, DFG SFB 1531, Project 456687919.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

