<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DO.MultiDGE — DO.MultiDGE • DOtools</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DO.MultiDGE — DO.MultiDGE"><meta name="description" content="Performs differential gene expression analysis using both single-cell and pseudo-bulk approaches across all annotated cell types.
The single-cell method uses Seurat's FindMarkers, while pseudo-bulk testing uses DESeq2 on aggregated expression profiles.
Outputs a merged data frame with DGE statistics from both methods per condition and cell type."><meta property="og:description" content="Performs differential gene expression analysis using both single-cell and pseudo-bulk approaches across all annotated cell types.
The single-cell method uses Seurat's FindMarkers, while pseudo-bulk testing uses DESeq2 on aggregated expression profiles.
Outputs a merged data frame with DGE statistics from both methods per condition and cell type."><script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.5.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adfct.html">Usage of functions</a></li>
    <li><a class="dropdown-item" href="../articles/cb.html">Remove ambient RNA with CellBender</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DO.MultiDGE</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarianoRuzJurado/DOtools/blob/master/R/utils_functions.R" class="external-link"><code>R/utils_functions.R</code></a></small>
      <div class="d-none name"><code>DO.MultiDGE.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Performs differential gene expression analysis using both single-cell and pseudo-bulk approaches across all annotated cell types.
The single-cell method uses Seurat's <code>FindMarkers</code>, while pseudo-bulk testing uses <code>DESeq2</code> on aggregated expression profiles.
Outputs a merged data frame with DGE statistics from both methods per condition and cell type.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DO.MultiDGE</span><span class="op">(</span></span>
<span>  <span class="va">Seu_obj</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>  method_sc <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>  group_by <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>  annotation_col <span class="op">=</span> <span class="st">"annotation"</span>,</span>
<span>  sample_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>  ident_ctrl <span class="op">=</span> <span class="st">"ctrl"</span>,</span>
<span>  min_pct <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  logfc_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  only_pos <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min_cells_group <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seu-obj">Seu_obj<a class="anchor" aria-label="anchor" href="#arg-seu-obj"></a></dt>
<dd><p>The seurat object</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Specified assay in Seurat object, default "RNA"</p></dd>


<dt id="arg-method-sc">method_sc<a class="anchor" aria-label="anchor" href="#arg-method-sc"></a></dt>
<dd><p>method to use for single cell DEG analysis, see FindMarkers from Seurat for options, default "wilcox"</p></dd>


<dt id="arg-group-by">group_by<a class="anchor" aria-label="anchor" href="#arg-group-by"></a></dt>
<dd><p>Column in meta data containing groups used for testing, default "condition"</p></dd>


<dt id="arg-annotation-col">annotation_col<a class="anchor" aria-label="anchor" href="#arg-annotation-col"></a></dt>
<dd><p>Column in meta data containing information of cell type annotation</p></dd>


<dt id="arg-sample-col">sample_col<a class="anchor" aria-label="anchor" href="#arg-sample-col"></a></dt>
<dd><p>Column in meta data containing information of sample annotation, default "orig.ident"</p></dd>


<dt id="arg-ident-ctrl">ident_ctrl<a class="anchor" aria-label="anchor" href="#arg-ident-ctrl"></a></dt>
<dd><p>Name of the condition in group_by to test against as ctrl, default "ctrl"</p></dd>


<dt id="arg-min-pct">min_pct<a class="anchor" aria-label="anchor" href="#arg-min-pct"></a></dt>
<dd><p>only test genes that are detected in a minimum fraction of min.pct cells in either of the two populations, default is 0</p></dd>


<dt id="arg-logfc-threshold">logfc_threshold<a class="anchor" aria-label="anchor" href="#arg-logfc-threshold"></a></dt>
<dd><p>Limit testing to genes which show, on average, at least X-fold difference (log-scale) between the two groups of cells, default is 0.</p></dd>


<dt id="arg-only-pos">only_pos<a class="anchor" aria-label="anchor" href="#arg-only-pos"></a></dt>
<dd><p>Only return positive markers, default FALSE</p></dd>


<dt id="arg-min-cells-group">min_cells_group<a class="anchor" aria-label="anchor" href="#arg-min-cells-group"></a></dt>
<dd><p>Minimum number of cells in one of the groups, default 3</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Additional arguments passed to FindMarkers function</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Dataframe containing statistics for each gene from the single cell and the Pseudobulk DGE approach.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mariano Ruz Jurado</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">sc_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sc_data.rds"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">DGE_result</span> <span class="op">&lt;-</span> <span class="fu">DO.MultiDGE</span><span class="op">(</span><span class="va">sc_data</span>,</span></span>
<span class="r-in"><span>                         sample_col <span class="op">=</span> <span class="st">"orig.ident"</span>,</span></span>
<span class="r-in"><span>                         method_sc <span class="op">=</span> <span class="st">"wilcox"</span>,</span></span>
<span class="r-in"><span>                         annotation_col <span class="op">=</span> <span class="st">"annotation"</span>,</span></span>
<span class="r-in"><span>                         ident_ctrl <span class="op">=</span> <span class="st">"healthy"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Names of identity class contain underscores ('_'), replacing with dashes ('-')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> <span style="color: #555555;">This message is displayed once every 8 hours.</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Centering and scaling data matrix</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:52 - Corrected annotation names in pseudo-bulk object by replacing '-' with '_'.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:52 - Starting DGE single cell method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:52 - Comparing disease with healthy in: Monocytes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> (default method for FindMarkers) please install the presto package</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> install.packages('devtools')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> devtools::install_github('immunogenomics/presto')</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> --------------------------------------------</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> After installation of presto, Seurat will automatically use the more </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> efficient implementation (no further action necessary).</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> This message will be shown once per session</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:52 - Comparing disease with healthy in: CD4_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:55 - Comparing disease with healthy in: CD8_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:55 - Skipping CD8_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:55 - Comparing disease with healthy in: NK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:56 - Comparing disease with healthy in: B_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:56 - Comparing disease with healthy in: pDC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Finished DGE single cell method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Starting DGE pseudo bulk method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: B_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping B_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: CD4_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping CD4_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: CD8_T_cells</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping CD8_T_cells since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: Monocytes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping Monocytes since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: NK</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping NK since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Comparing disease with healthy in: pDC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Skipping pDC since one comparison has fewer than 3 cells!</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - Finished DGE pseudo bulk method analysis</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2025-07-08 11:49:57 - DGE pseudo bulk result is empty...</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

