<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DOtools: A Toolkit for scRNA Data Analysis — DOtools • DOtools</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DOtools: A Toolkit for scRNA Data Analysis — DOtools"><meta name="description" content="The `DOtools` package provides a set of functions for advanced data processing, visualisation, and statistical analysis in Seurat objects. It includes functions for cell-type prediction, reclustering, creating polished UMAP plots, subsetting Seurat objects, and various statistical analyses like Wilcoxon tests and SEM graphs."><meta property="og:description" content="The `DOtools` package provides a set of functions for advanced data processing, visualisation, and statistical analysis in Seurat objects. It includes functions for cell-type prediction, reclustering, creating polished UMAP plots, subsetting Seurat objects, and various statistical analyses like Wilcoxon tests and SEM graphs."><script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/adfct.html">Usage of functions</a></li>
    <li><a class="dropdown-item" href="../articles/cb.html">Remove ambient RNA with CellBender</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DOtools: A Toolkit for scRNA Data Analysis</h1>

      <div class="d-none name"><code>DOtools.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The `DOtools` package provides a set of functions for advanced data processing, visualisation, and statistical analysis in Seurat objects. It includes functions for cell-type prediction, reclustering, creating polished UMAP plots, subsetting Seurat objects, and various statistical analyses like Wilcoxon tests and SEM graphs.</p>
    </div>


    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This package includes the following functions:</p><ul><li><p><code><a href="DO.BoxPlot.html">DO.BoxPlot</a></code>: A function for creating box plots with Wilcoxon test results.</p></li>
<li><p><code><a href="DO.CellTypist.html">DO.CellTypist</a></code>: A function for running CellTypist on Seurat objects to predict cell types.</p></li>
<li><p><code><a href="DO.DietSeurat.html">DO.DietSeurat</a></code>: A function for diet-based analysis of Seurat objects.</p></li>
<li><p><code><a href="DO.Dotplot.html">DO.Dotplot</a></code>: A function for creating dot plots for visualizing gene expression.</p></li>
<li><p><code><a href="DO.FullRecluster.html">DO.FullRecluster</a></code>: A function for fine-grained reclustering of Seurat objects.</p></li>
<li><p><code><a href="DO.BarplotClustert.html">DO.BarplotClustert</a></code>: A function for generating mean and SEM graphs for cluster-based analysis with t-tests.</p></li>
<li><p><code><a href="DO.BarplotWilcox.html">DO.BarplotWilcox</a></code>: A function for generating mean and SEM graphs with Wilcoxon test results.</p></li>
<li><p><code><a href="DO.Subset.html">DO.Subset</a></code>: A function for subsetting Seurat objects based on metadata.</p></li>
<li><p><code><a href="DO.UMAP.html">DO.UMAP</a></code>: A function for creating polished UMAP plots using either DimPlot or FeaturePlot.</p></li>
<li><p><code><a href="DO.VlnPlot.html">DO.VlnPlot</a></code>: A function for generating violin plots with Wilcoxon test results.</p></li>
<li><p><code><a href="DO.CellComposition.html">DO.CellComposition</a></code>: A function for visualizing and statistically analyzing cell-type composition changes across conditions using the Scanpro Python package, with support for bootstrapping, proportion plots, and customizable output.</p></li>
<li><p><code><a href="DO.Import.html">DO.Import</a></code>: A function for building a merged Seurat object from 10x software output, or directly from provided tables.</p></li>
<li><p><code><a href="DO.CellBender.html">DO.CellBender</a></code>: A function for running CellBender in a virtual conda env with provided raw count h5 files.</p></li>
<li><p><code><a href="DO.SplitBarGSEA.html">DO.SplitBarGSEA</a></code>: A function for viusalizing GSEA result from a provided df from e.g. metascape</p></li>
<li><p><code><a href="DO.scVI.html">DO.scVI</a></code>: A function for running the scVI Integration implemented in scvi-tools.</p></li>
<li><p><code><a href="DO.TransferLabel.html">DO.TransferLabel</a></code>: A function for transfering annotation from a subseted object to the original seurat object.</p></li>
<li><p><code><a href="DO.PyEnv.html">DO.PyEnv</a></code>: A function for creating a conda envrionment holding all python packages needed for some functions.</p></li>
<li><p><code><a href="DO.Correlation.html">DO.Correlation</a></code>: A function for creating a correlation plot between provided samples in the category specified.</p></li>
<li><p><code><a href="DO.Heatmap.html">DO.Heatmap</a></code>: A function for generating Heat maps on gene expression data.</p></li>
<li><p><code><a href="DO.MultiDGE.html">DO.MultiDGE</a></code>: A function for calculating DEGs on a single cell and speudo bulk level.</p></li>
<li><p><code>dot-Do.BarcodeRanks</code>: A function for estimating the number of expected cells and droplets.</p></li>
<li><p><code>dot-QC.Vlnplot</code>: A function for estimating the number of expected cells and droplets.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mariano Ruz Jurado, David Rodriguez Morales</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="DO.BoxPlot.html">DO.BoxPlot</a></code>, <code><a href="DO.CellTypist.html">DO.CellTypist</a></code>, <code><a href="DO.DietSeurat.html">DO.DietSeurat</a></code>, <code><a href="DO.Dotplot.html">DO.Dotplot</a></code>,
    <code><a href="DO.FullRecluster.html">DO.FullRecluster</a></code>, <code><a href="DO.BarplotClustert.html">DO.BarplotClustert</a></code>, <code><a href="DO.BarplotWilcox.html">DO.BarplotWilcox</a></code>,
    <code><a href="DO.Subset.html">DO.Subset</a></code>, <code><a href="DO.UMAP.html">DO.UMAP</a></code>, <code><a href="DO.VlnPlot.html">DO.VlnPlot</a></code>, <code><a href="DO.Import.html">DO.Import</a></code>, <code><a href="DO.CellBender.html">DO.CellBender</a></code>,
    <code><a href="DO.SplitBarGSEA.html">DO.SplitBarGSEA</a></code>, <code><a href="DO.scVI.html">DO.scVI</a></code>, <code><a href="DO.TransferLabel.html">DO.TransferLabel</a></code>,<code><a href="DO.Heatmap.html">DO.Heatmap</a></code>, <code><a href="DO.PyEnv.html">DO.PyEnv</a></code>, <code><a href="DO.Correlation.html">DO.Correlation</a></code>,
    <code><a href="DO.MultiDGE.html">DO.MultiDGE</a></code>,<code><a href="DO.TransferLabel.html">DO.TransferLabel</a></code>, <code>dot-Do.BarcodeRanks</code>, <code>dot-QC.Vlnplot</code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

