<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>DO.Integration SCE object integration directly — DO.Integration • DOtools</title><!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="DO.Integration SCE object integration directly — DO.Integration"><meta name="description" content="Integrates single-cell RNA-seq data directly from
SingleCellExperiment or Seurat objects. Supports detection of variable genes
, scaling, PCA, neighbor graph construction, clustering, and UMAP embedding,
with multiple integration methods."><meta property="og:description" content="Integrates single-cell RNA-seq data directly from
SingleCellExperiment or Seurat objects. Supports detection of variable genes
, scaling, PCA, neighbor graph construction, clustering, and UMAP embedding,
with multiple integration methods."><script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.7</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch"><li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>DO.Integration SCE object integration directly</h1>
      <small class="dont-index">Source: <a href="https://github.com/MarianoRuzJurado/DOtools/blob/HEAD/R/DO.Integration.R" class="external-link"><code>R/DO.Integration.R</code></a></small>
      <div class="d-none name"><code>DO.Integration.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Integrates single-cell RNA-seq data directly from
SingleCellExperiment or Seurat objects. Supports detection of variable genes
, scaling, PCA, neighbor graph construction, clustering, and UMAP embedding,
with multiple integration methods.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">DO.Integration</span><span class="op">(</span></span>
<span>  <span class="va">sce_object</span>,</span>
<span>  split_key <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>  HVG <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pca <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  neighbors <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  neighbors_dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  clusters <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  clusters_res <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  clusters_algorithm <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  umap <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  umap_key <span class="op">=</span> <span class="st">"UMAP"</span>,</span>
<span>  umap_dim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fl">50</span><span class="op">)</span>,</span>
<span>  integration_method <span class="op">=</span> <span class="st">"CCAIntegration"</span>,</span>
<span>  selection_method <span class="op">=</span> <span class="st">"vst"</span>,</span>
<span>  loess_span <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  clip_max <span class="op">=</span> <span class="st">"auto"</span>,</span>
<span>  num_bin <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  binning_method <span class="op">=</span> <span class="st">"equal_width"</span>,</span>
<span>  scale_max <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  pca_key <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>  integration_key <span class="op">=</span> <span class="st">"INTEGRATED.CCA"</span>,</span>
<span>  npcs <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-sce-object">sce_object<a class="anchor" aria-label="anchor" href="#arg-sce-object"></a></dt>
<dd><p>Seurat or SCE Object</p></dd>


<dt id="arg-split-key">split_key<a class="anchor" aria-label="anchor" href="#arg-split-key"></a></dt>
<dd><p>Character. Column in meta data to split the samples by,
default orig.ident</p></dd>


<dt id="arg-hvg">HVG<a class="anchor" aria-label="anchor" href="#arg-hvg"></a></dt>
<dd><p>Logical. Perform detection of highly variable genes</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Logical. Perform scaling of the expression data</p></dd>


<dt id="arg-pca">pca<a class="anchor" aria-label="anchor" href="#arg-pca"></a></dt>
<dd><p>Logical. Perform principal component analysis</p></dd>


<dt id="arg-neighbors">neighbors<a class="anchor" aria-label="anchor" href="#arg-neighbors"></a></dt>
<dd><p>Logical. Perform Nearest-neighbor graph after integration</p></dd>


<dt id="arg-neighbors-dim">neighbors_dim<a class="anchor" aria-label="anchor" href="#arg-neighbors-dim"></a></dt>
<dd><p>Numeric range. Dimensions of reduction to use as input</p></dd>


<dt id="arg-clusters">clusters<a class="anchor" aria-label="anchor" href="#arg-clusters"></a></dt>
<dd><p>Logical. Perform clustering of cells</p></dd>


<dt id="arg-clusters-res">clusters_res<a class="anchor" aria-label="anchor" href="#arg-clusters-res"></a></dt>
<dd><p>Numeric. Value of the resolution parameter, use a value
above (below) 1.0 if you want to obtain a larger (smaller) number of
communities.</p></dd>


<dt id="arg-clusters-algorithm">clusters_algorithm<a class="anchor" aria-label="anchor" href="#arg-clusters-algorithm"></a></dt>
<dd><p>Numeric. Define the algorithm for clustering,
default 4 for "Leiden"</p></dd>


<dt id="arg-umap">umap<a class="anchor" aria-label="anchor" href="#arg-umap"></a></dt>
<dd><p>Logical. Runs the Uniform Manifold Approximation and Projection</p></dd>


<dt id="arg-umap-key">umap_key<a class="anchor" aria-label="anchor" href="#arg-umap-key"></a></dt>
<dd><p>Character name for</p></dd>


<dt id="arg-umap-dim">umap_dim<a class="anchor" aria-label="anchor" href="#arg-umap-dim"></a></dt>
<dd><p>Numeric range. Which dimensions to use as input features</p></dd>


<dt id="arg-integration-method">integration_method<a class="anchor" aria-label="anchor" href="#arg-integration-method"></a></dt>
<dd><p>Character. Define the integration method, please
check what versions are supported in Seurat::IntegrateLayers function</p></dd>


<dt id="arg-selection-method">selection_method<a class="anchor" aria-label="anchor" href="#arg-selection-method"></a></dt>
<dd><p>Character. Default "vst". Options: "mean.var.plot",
"dispersion"</p></dd>


<dt id="arg-loess-span">loess_span<a class="anchor" aria-label="anchor" href="#arg-loess-span"></a></dt>
<dd><p>Numeric. Loess span parameter used when fitting the
variance-mean relationship</p></dd>


<dt id="arg-clip-max">clip_max<a class="anchor" aria-label="anchor" href="#arg-clip-max"></a></dt>
<dd><p>Character. After standardization values larger than clip.max
will be set to clip.max; default is 'auto' which sets this value to the
square root of the number of cells</p></dd>


<dt id="arg-num-bin">num_bin<a class="anchor" aria-label="anchor" href="#arg-num-bin"></a></dt>
<dd><p>Numeric. Total number of bins to use in the scaled analysis
(default is 20)</p></dd>


<dt id="arg-binning-method">binning_method<a class="anchor" aria-label="anchor" href="#arg-binning-method"></a></dt>
<dd><p>Character. “equal_width”: each bin is of equal width
along the x-axis (default). Options: “equal_frequency”:</p></dd>


<dt id="arg-scale-max">scale_max<a class="anchor" aria-label="anchor" href="#arg-scale-max"></a></dt>
<dd><p>Numeric. Max value to return for scaled data. The default
is 10.</p></dd>


<dt id="arg-pca-key">pca_key<a class="anchor" aria-label="anchor" href="#arg-pca-key"></a></dt>
<dd><p>Character. Key name to save the pca result in</p></dd>


<dt id="arg-integration-key">integration_key<a class="anchor" aria-label="anchor" href="#arg-integration-key"></a></dt>
<dd><p>Character. Key name to save the integration result in</p></dd>


<dt id="arg-npcs">npcs<a class="anchor" aria-label="anchor" href="#arg-npcs"></a></dt>
<dd><p>Numeric. Total Number of PCs to compute and store (50 by default)</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Logical. Verbosity for all functions</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>integrated sce/seurat object</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Mariano Ruz Jurado</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">sce_data</span> <span class="op">&lt;-</span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce_data.rds"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">DO.Integration</span><span class="op">(</span></span></span>
<span class="r-in"><span>    sce_object <span class="op">=</span> <span class="va">sce_data</span>,</span></span>
<span class="r-in"><span>    split_key <span class="op">=</span> <span class="st">"orig.ident"</span>,</span></span>
<span class="r-in"><span>    HVG <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    scale <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    pca <span class="op">=</span> <span class="cn">TRUE</span>,</span></span>
<span class="r-in"><span>    integration_method <span class="op">=</span> <span class="st">"CCAIntegration"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:25:55 - Splitting object for integration with CCAIntegration by orig.ident</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:25:55 - Calculating highly variable genes</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:25:56 - Scaling object</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:25:56 - Running pca, saved in key: PCA</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Splitting ‘counts’, ‘data’ layers. Not splitting ‘scale.data’. If you would like to split other layers, set in `layers` argument.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:25:57 - Running integration, saved in key: INTEGRATED.CCA</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:26:00 - Running Nearest-neighbor graph construction</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:26:01 - Running cluster detection</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> 2025-10-02 15:26:02 - Creating UMAP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SingleCellExperiment </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> dim: 800 2807 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> metadata(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> assays(3): counts logcounts scaledata</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rownames(800): HES4 ISG15 ... SERPINA9 DSG2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> rowData names(0):</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colnames(2807): AAACCCAGTGCATTTG-1_1 AAACCCATCTCAACGA-1_1 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   TTTGGAGCAACTGGTT-1_2 TTTGGAGGTTACCTGA-1_2</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> colData names(15): orig.ident nCount_RNA ... annotation_recluster</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   leiden0.3</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> reducedDimNames(3): PCA INTEGRATED.CCA UMAP</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> mainExpName: RNA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> altExpNames(0):</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John, DFG SFB 1366, Project B04, DFG SFB 1531, Project 456687919.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

