<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quality control of sc/snRNA-seq • DOtools</title>
<!-- mathjax math --><script src="https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/tex-mml-chtml.js" integrity="sha256-MASABpB4tYktI2Oitl4t+78w/lyA+D7b/s9GEP0JOGI=" crossorigin="anonymous"></script><script>
  window.MathJax = {
    chtml: {
      fontURL: "https://cdn.jsdelivr.net/npm/mathjax@3.2.2/es5/output/chtml/fonts/woff-v2"
    }
  };
</script><script src="../lightswitch.js"></script><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quality control of sc/snRNA-seq">
<script defer data-domain="pkgdown.r-lib.org,all.tidyverse.org" src="https://plausible.io/js/plausible.js"></script>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top " aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">DOtools</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Released version">0.99.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/DOtools.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/adfct.html">Usage of functions</a></li>
    <li><a class="dropdown-item" href="../articles/cb.html">Remove ambient RNA with CellBender</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/MarianoRuzJurado/DOtools/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-lightswitch" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Light switch"><span class="fa fa-sun"></span></button>
  <ul class="dropdown-menu dropdown-menu-end" aria-labelledby="dropdown-lightswitch">
<li><button class="dropdown-item" data-bs-theme-value="light"><span class="fa fa-sun"></span> Light</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="dark"><span class="fa fa-moon"></span> Dark</button></li>
    <li><button class="dropdown-item" data-bs-theme-value="auto"><span class="fa fa-adjust"></span> Auto</button></li>
  </ul>
</li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">



<script src="DOtools_files/kePrint-0.0.1/kePrint.js"></script><link href="DOtools_files/lightable-0.0.1/lightable.css" rel="stylesheet">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quality control of sc/snRNA-seq</h1>
                        <h4 data-toc-skip class="author">Mariano Ruz
Jurado</h4>
            <address class="author_afil">
      Goethe University<br><small class="dont-index">Source: <a href="https://github.com/MarianoRuzJurado/DOtools/blob/master/vignettes/DOtools.Rmd" class="external-link"><code>vignettes/DOtools.Rmd</code></a></small>
      <div class="d-none name"><code>DOtools.Rmd</code></div>
    </address>
</div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.21/DOtools" class="external-link">DOtools</a></em>
is an R package distributed as part of the Bioconductor project. To
install the package, start R and enter:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span> <span class="co"># WORK iN PROGRESS</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"DOtools"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively, you can instead install the latest development version
from <a href="https://github.com/" class="external-link"><em>GitHub</em></a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"MarianoRuzJurado/DOtools"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="usage">Usage<a class="anchor" aria-label="anchor" href="#usage"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.21/DOtools" class="external-link">DOtools</a></em>
contains different functions for processing and visualizing gene
expression in scRNA/snRNA experiments:</p>
<p>In this vignette we showcase how to use the functions with public
available data.</p>
<div class="section level3">
<h3 id="libraries">Libraries<a class="anchor" aria-label="anchor" href="#libraries"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.21/DOtools" class="external-link">DOtools</a></em>
can be imported as:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://marianoruzjurado.github.io/DOtools/">DOtools</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Additional packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://had.co.nz/plyr" class="external-link">plyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">enrichR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://haozhu233.github.io/kableExtra/" class="external-link">kableExtra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h3>
<p><em><a href="https://bioconductor.org/packages/3.21/DOtools" class="external-link">DOtools</a></em>
The <code><a href="../reference/DO.Import.html">DO.Import()</a></code> function provides a streamlined pipeline
for performing quality control on single-cell or single-nucleus RNA
sequencing (sc/snRNA-seq) data. It takes as input a list of .h5 files
generated by e.g. CellRanger or STARsolo, along with sample names and
metadata.</p>
<p>During preprocessing, low-quality genes and cells are filtered out
based on specified thresholds. Genes expressed in fewer than five cells
are removed. Cells are filtered according to mitochondrial gene content,
number of detected genes, total UMI counts, and potential doublets. The
function supports doublet detection using <em><a href="https://bioconductor.org/packages/3.21/scDblFinder" class="external-link">scDblFinder</a></em>.
Thresholds for mitochondrial content (e.g., 5% for scRNA-seq and 3% for
snRNA-seq), gene counts, and UMI counts can be defined to tailor the
filtering.</p>
<p>After filtering, samples are merged into one <a href="https://cran.r-project.org/package=Seurat" class="external-link"><em>Seurat</em></a> or
<a href="https://cran.r-project.org/package=SingleCellExperiment" class="external-link"><em>SingleCellExperiment</em></a>
object, followed by log-normalisation, scaling, and the identification
of highly variable genes. To help assess the effect of quality control,
violin plots showing distributions of key metrics before and after
filtering are automatically generated and saved alongside the input
files. A summary of removed genes and cells is also recorded.</p>
<p>To show how the quality control works, we are going to use a public
dataset from 10X from human blood of healthy and donors with a malignant
tumor:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">base</span> <span class="op">&lt;-</span> <span class="fu">DOtools</span><span class="fu">:::</span><span class="fu"><a href="../reference/dot-example_10x.html">.example_10x</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">paths</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"healthy/outs/filtered_feature_bc_matrix.h5"</span><span class="op">)</span>,</span>
<span>         <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"disease/outs/filtered_feature_bc_matrix.h5"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SCE_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.Import.html">DO.Import</a></span><span class="op">(</span>pathways <span class="op">=</span> <span class="va">paths</span>,</span>
<span>                     ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"healthy-1"</span>, <span class="st">"disease-1"</span><span class="op">)</span>,</span>
<span>                     DeleteDoublets <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                     cut_mt <span class="op">=</span> <span class="fl">.05</span>,</span>
<span>                     min_counts <span class="op">=</span> <span class="fl">500</span>,</span>
<span>                     min_genes <span class="op">=</span> <span class="fl">300</span>,</span>
<span>                     high_quantile <span class="op">=</span> <span class="fl">.95</span>,</span>
<span>                     Seurat<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span> <span class="co"># Set to TRUE for Seurat object</span></span></code></pre></div>
<p>We can now check the quality before introducing filterings:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prefilterplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="st">"prefilterplots-1.png"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span></span>
<span><span class="va">pQC1</span> <span class="op">&lt;-</span> <span class="fu">magick</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="va">prefilterplots</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pQC1</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/prefilterplots-1.png" width="100%" style="display: block; margin: auto;"><br> And after:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">postfilterplots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"figures"</span>, <span class="st">"postfilterplots-1.png"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span></span>
<span><span class="va">pQC2</span> <span class="op">&lt;-</span> <span class="fu">magick</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read</a></span><span class="op">(</span><span class="va">postfilterplots</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">pQC2</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/postfilterplots-1.png" width="100%" style="display: block; margin: auto;">
We observed that most cells were removed due to increased mitochondrial
content. Depending on the experimental design, the mitochondrial content
threshold can be adjusted to retain a higher number of cells, if
minimizing cell loss is of relevance.</p>
<p><br> The DOtools package provides a slim object of this data set.
Please feel free, to use the one created from DO.Import for prettier
results or this slim downed version. We can observe how similar the
samples are through running a correlation analysis.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Making sure we have a save folder</span></span>
<span><span class="va">base</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"my_tempdir_"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">base</span><span class="op">)</span></span>
<span>  </span>
<span><span class="va">SCE_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"sce_data.rds"</span>, package <span class="op">=</span> <span class="st">"DOtools"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/DO.Correlation.html">DO.Correlation</a></span><span class="op">(</span><span class="va">SCE_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `PackageCheck()` was deprecated in SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use `rlang::check_installed()` instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; Warning: The `slot` argument of `SetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">#&gt;   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;">generated.</span></span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">fill</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">fill</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/heatmap-1.png" width="576"></p>
</div>
<div class="section level3">
<h3 id="data-integration">Data integration<a class="anchor" aria-label="anchor" href="#data-integration"></a>
</h3>
<p>After quality control the prefered integration method can be chosen
within Seurat’s <code>IntegrateLayers</code> function. Additionally, we
implemented a new wrapper function for the scVI integration from the <a href="https://pypi.org/project/scvi-tools/" class="external-link">scvi-tools</a> package.
After the integration completes, we run the Leiden algorithm to find
clusters and generate UMAP embeddings.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">### ADDITIONAL STEPS FOR SCE OBJECTS, THESE STEPS WILL BE PERFORMED AUTOMATICALLY FOR SEURAT OBJECTS IN DO.IMPORT</span></span>
<span><span class="co">### SKIP IF YOU USE SEURAT OBJECT FROM DO.IMPORT() </span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/as.Seurat.html" class="external-link">as.Seurat</a></span><span class="op">(</span><span class="va">SCE_obj</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">Seu_obj</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding variable features for layer counts.healthy-1</span></span>
<span><span class="co">#&gt; Finding variable features for layer counts.disease-1</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">Seu_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">Seu_obj</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span>, reduction.name <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">[[</span><span class="st">"RNA"</span><span class="op">]</span><span class="op">]</span>, f <span class="op">=</span> <span class="va">Seu_obj</span><span class="op">$</span><span class="va">orig.ident</span><span class="op">)</span></span>
<span><span class="co">#&gt; Splitting 'counts', 'data' layers. Not splitting 'scale.data'. If you would like to split other layers, set in `layers` argument.</span></span>
<span><span class="co">#####</span></span>
<span></span>
<span><span class="co">#Integration through Seurat</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/IntegrateLayers.html" class="external-link">IntegrateLayers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">Seu_obj</span>,</span>
<span>                           method <span class="op">=</span> <span class="va">CCAIntegration</span>,</span>
<span>                           orig.reduction <span class="op">=</span> <span class="st">"PCA"</span>,</span>
<span>                           new.reduction <span class="op">=</span> <span class="st">"INTEGRATED.CCA"</span>,</span>
<span>                           verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding all pairwise anchors</span></span>
<span><span class="co">#&gt; Running CCA</span></span>
<span><span class="co">#&gt; Merging objects</span></span>
<span><span class="co">#&gt; Finding neighborhoods</span></span>
<span><span class="co">#&gt; Finding anchors</span></span>
<span><span class="co">#&gt;  Found 2607 anchors</span></span>
<span><span class="co">#&gt; Merging dataset 2 into 1</span></span>
<span><span class="co">#&gt; Extracting anchors for merged samples</span></span>
<span><span class="co">#&gt; Finding integration vectors</span></span>
<span><span class="co">#&gt; Finding integration vector weights</span></span>
<span><span class="co">#&gt; Integrating data</span></span>
<span></span>
<span><span class="co">#After Integration we join the layers </span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/SplitLayers.html" class="external-link">JoinLayers</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#(Optional) convert back to SCE object if you prefer that data class</span></span>
<span><span class="va">SCE_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html" class="external-link">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#(Optional) Integration with scVI-Model</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.scVI.html">DO.scVI</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">Seu_obj</span>,</span>
<span>                   batch_key <span class="op">=</span><span class="st">"orig.ident"</span>,</span>
<span>                   layer_counts <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>                   layer_logcounts <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span></span></code></pre></div>
<p>After the integration finished, both corrected expression matrices
can be found saved in the Seurat object and can be used for cluster
calculations and UMAP projections. In this case, we will continue with
Seuratv5 CCA Integration method.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">Seu_obj</span>, reduction <span class="op">=</span> <span class="st">"INTEGRATED.CCA"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span> <span class="co">#change dims accordingly</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">Seu_obj</span>, resolution <span class="op">=</span> <span class="fl">0.3</span>, algorithm <span class="op">=</span> <span class="fl">4</span>, random.seed <span class="op">=</span> <span class="fl">1234</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">Seu_obj</span>, reduction <span class="op">=</span> <span class="st">"INTEGRATED.CCA"</span>, reduction.name <span class="op">=</span> <span class="st">"UMAP"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; 07:33:41 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 07:33:41 Read 2807 rows and found 50 numeric columns</span></span>
<span><span class="co">#&gt; 07:33:41 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 07:33:41 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 07:33:41 Writing NN index file to temp file /tmp/Rtmp3adQTx/file26f22c8c860a</span></span>
<span><span class="co">#&gt; 07:33:41 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 07:33:42 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 07:33:43 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 07:33:45 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 07:33:45 Commencing optimization for 500 epochs, with 130642 positive edges</span></span>
<span><span class="co">#&gt; 07:33:45 Using rng type: pcg</span></span>
<span><span class="co">#&gt; 07:33:48 Optimization finished</span></span>
<span></span>
<span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Clustering%20and%20UMAP-1.png" width="700"></p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"condition"</span>, legend.position <span class="op">=</span> <span class="st">"right"</span>, label <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Clustering%20and%20UMAP-2.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="semi-automatic-annotation-with-celltypist">Semi-automatic annotation with Celltypist<a class="anchor" aria-label="anchor" href="#semi-automatic-annotation-with-celltypist"></a>
</h3>
<p>Next up, we implemented a wrapper around the semi-automatic
annotation tool <a href="https://pypi.org/project/celltypist/" class="external-link"><em>celltypist</em></a>. It
will annotate the defined clusters based on the
<code>Adult_COVID19_PBMC.pkl</code> model.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.CellTypist.html">DO.CellTypist</a></span><span class="op">(</span><span class="va">Seu_obj</span>,</span>
<span>                         modelName <span class="op">=</span> <span class="st">"Healthy_COVID19_PBMC.pkl"</span>, </span>
<span>                         runCelltypistUpdate<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                         over_clustering <span class="op">=</span> <span class="st">"seurat_clusters"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-07-26 07:33:49 - Running celltypist using model: Healthy_COVID19_PBMC.pkl</span></span>
<span><span class="co">#&gt; 2025-07-26 07:33:49 - Saving celltypist results to temporary folder: /tmp/Rtmp3adQTx/file26f272714652</span></span>
<span><span class="co">#&gt; 2025-07-26 07:34:58 - Creating probality plot</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/annotation-1.png" width="700"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"autoAnnot"</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/annotation-2.png" width="700"></p>
<p>The semi-automatic annotation is a good estimate of the cell types in
your object. But you should always manually validate the findings of the
model. You can manually define a set of marker genes for the cell
population or check the most preeminent genes per cluster by using
Seurat’s <code>FindAllMarkers</code> function.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#pick top 5 per cluster</span></span>
<span><span class="va">annotation_Markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">Seu_obj</span>,</span>
<span>                                     assay <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                                     group.by <span class="op">=</span> <span class="st">"autoAnnot"</span>,</span>
<span>                                     min.pct <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                                     logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster CD83_CD14_mono</span></span>
<span><span class="co">#&gt; For a (much!) faster implementation of the Wilcoxon Rank Sum Test,</span></span>
<span><span class="co">#&gt; (default method for FindMarkers) please install the presto package</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; install.packages('devtools')</span></span>
<span><span class="co">#&gt; devtools::install_github('immunogenomics/presto')</span></span>
<span><span class="co">#&gt; --------------------------------------------</span></span>
<span><span class="co">#&gt; After installation of presto, Seurat will automatically use the more </span></span>
<span><span class="co">#&gt; efficient implementation (no further action necessary).</span></span>
<span><span class="co">#&gt; This message will be shown once per session</span></span>
<span><span class="co">#&gt; Calculating cluster CD4.Naive</span></span>
<span><span class="co">#&gt; Calculating cluster CD8.Naive</span></span>
<span><span class="co">#&gt; Calculating cluster NK_16hi</span></span>
<span><span class="co">#&gt; Calculating cluster CD8.EM</span></span>
<span><span class="co">#&gt; Calculating cluster B_naive</span></span>
<span><span class="co">#&gt; Calculating cluster pDC</span></span>
<span></span>
<span><span class="va">annotation_Markers</span> <span class="op">&lt;-</span> <span class="va">annotation_Markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/plyr/man/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">avg_log2FC</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>              </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="va">gene</span>, .keep_all <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span>       </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice_head</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>   </span>
<span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.Dotplot.html">DO.Dotplot</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">Seu_obj</span>,</span>
<span>                 Feature <span class="op">=</span> <span class="va">annotation_Markers</span>,</span>
<span>                 group.by.x <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>                 plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                 annotation_x <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 point_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                 annotation_x_rev <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 textSize <span class="op">=</span> <span class="fl">14</span>,</span>
<span>                 hjust <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                 vjust <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                 textRot <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                 segWidth <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                 lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">size</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">size</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Manual%20annotation-1.png" width="1152"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#manual set of markers</span></span>
<span><span class="va">annotation_Markers</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cluster <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ImmuneCells"</span>,</span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"B_cells"</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"T_cells"</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"NK"</span>, <span class="fl">2</span><span class="op">)</span>,</span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"Myeloid"</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>                                            <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"pDC"</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"PTPRC"</span>, <span class="st">"CD79A"</span>, <span class="st">"BANK1"</span>, <span class="st">"MS4A1"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD4"</span>, <span class="st">"IL7R"</span>, <span class="st">"NKG7"</span>, </span>
<span>                                          <span class="st">"KLRD1"</span>,<span class="st">"CD68"</span>, <span class="st">"CD14"</span>,<span class="st">"ITGAM"</span>, <span class="st">"LILRA4"</span>, <span class="st">"CLEC4C"</span>, <span class="st">"LRRC26"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.Dotplot.html">DO.Dotplot</a></span><span class="op">(</span>sce_object <span class="op">=</span> <span class="va">Seu_obj</span>,</span>
<span>                 Feature <span class="op">=</span> <span class="va">annotation_Markers</span>,</span>
<span>                 group.by.x <span class="op">=</span> <span class="st">"seurat_clusters"</span>,</span>
<span>                 plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>                 annotation_x <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 point_stroke <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                 annotation_x_rev <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                 textSize <span class="op">=</span> <span class="fl">14</span>,</span>
<span>                 hjust <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                 vjust <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                 textRot <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                 segWidth <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                 lwd <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for <span style="color: #00BB00;">size</span> is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for <span style="color: #00BB00;">size</span>, which will replace the existing scale.</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Manual%20annotation-2.png" width="1152"></p>
<p>The manual markers for the immune cells show an agreement for the
annotation therefore we can continue with it after some minor
adjustments</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Seu_obj</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/revalue.html" class="external-link">revalue</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">$</span><span class="va">seurat_clusters</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`1` <span class="op">=</span> <span class="st">"T_cells"</span>,</span>
<span>                                                               `2` <span class="op">=</span> <span class="st">"T_cells"</span>,</span>
<span>                                                               `3` <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>                                                               `4` <span class="op">=</span> <span class="st">"B_cells"</span>,</span>
<span>                                                               `5` <span class="op">=</span> <span class="st">"Monocytes"</span>,</span>
<span>                                                               `6` <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>                                                               `7` <span class="op">=</span> <span class="st">"T_cells"</span>,</span>
<span>                                                               `8` <span class="op">=</span> <span class="st">"pDC"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"annotation"</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/renaming-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="cell-composition">Cell composition<a class="anchor" aria-label="anchor" href="#cell-composition"></a>
</h3>
<p>After the identification of the celltype populations, we can also
evaluate if there are significant changes in these populations in the
healthy and diseased condition using a wrapper function around the
python tool <a href="https://pypi.org/project/scanpro/" class="external-link"><em>scanpro</em></a>.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/DO.CellComposition.html">DO.CellComposition</a></span><span class="op">(</span><span class="va">Seu_obj</span>,</span>
<span>                   assay_normalized <span class="op">=</span> <span class="st">"RNA"</span>,</span>
<span>                   cluster_column <span class="op">=</span> <span class="st">"annotation"</span>,</span>
<span>                   sample_column <span class="op">=</span> <span class="st">"orig.ident"</span>,</span>
<span>                   condition_column <span class="op">=</span> <span class="st">"condition"</span>,</span>
<span>                   transform_method <span class="op">=</span> <span class="st">"arcsin"</span>,</span>
<span>                   n_reps <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:07 - Bootstrapping method activated with 3 simulated replicates!</span></span>
<span><span class="co">#&gt; .</span></span>
<span><span class="co">#&gt; Using orig.ident, condition as id variables</span></span>
<span><span class="co">#&gt; Using condition as id variables</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/cell%20populations-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="reclustering-of-cell-populations">Reclustering of cell populations<a class="anchor" aria-label="anchor" href="#reclustering-of-cell-populations"></a>
</h3>
<p>Subpopulations can be tricky to find, therefore it is always a good
practice to perform a reclustering of a given cell populations, if we
are interested in a specific set of cells in a population. Here for
example in the T cells. We will identify the subpopulations and then
markers defining them.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.FullRecluster.html">DO.FullRecluster</a></span><span class="op">(</span><span class="va">Seu_obj</span>, over_clustering <span class="op">=</span> <span class="st">"annotation"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"annotation_recluster"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Recluster-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">T_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.Subset.html">DO.Subset</a></span><span class="op">(</span><span class="va">Seu_obj</span>,</span>
<span>                     ident <span class="op">=</span> <span class="st">"annotation_recluster"</span>,</span>
<span>                     ident_name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"T_cells"</span>, <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">$</span><span class="va">annotation_recluster</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:26 - Specified 'ident_name': expecting a categorical variable.</span></span>
<span></span>
<span><span class="va">T_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html" class="external-link">FindAllMarkers</a></span><span class="op">(</span><span class="va">T_cells</span>, group.by <span class="op">=</span> <span class="st">"annotation_recluster"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster T_cells_1</span></span>
<span><span class="co">#&gt; Calculating cluster T_cells_3</span></span>
<span><span class="co">#&gt; Calculating cluster T_cells_2</span></span>
<span></span>
<span><span class="va">T_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.CellTypist.html">DO.CellTypist</a></span><span class="op">(</span><span class="va">T_cells</span>,</span>
<span>                         modelName <span class="op">=</span> <span class="st">"Healthy_COVID19_PBMC.pkl"</span>,</span>
<span>                         runCelltypistUpdate<span class="op">=</span><span class="cn">FALSE</span>,</span>
<span>                         over_clustering <span class="op">=</span> <span class="st">"annotation_recluster"</span>,</span>
<span>                         SeuV5 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:31 - Running celltypist using model: Healthy_COVID19_PBMC.pkl</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:31 - Saving celltypist results to temporary folder: /tmp/Rtmp3adQTx/file26f22eb5beae</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:45 - Creating probality plot</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/Recluster-2.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">T_cells</span><span class="op">$</span><span class="va">annotation</span> <span class="op">&lt;-</span> <span class="fu">plyr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plyr/man/revalue.html" class="external-link">revalue</a></span><span class="op">(</span><span class="va">T_cells</span><span class="op">$</span><span class="va">annotation_recluster</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>`T_cells_1` <span class="op">=</span> <span class="st">"CD4_T_cells"</span>,</span>
<span>                                                                    `T_cells_2` <span class="op">=</span> <span class="st">"CD4_T_cells"</span>,</span>
<span>                                                                    `T_cells_3` <span class="op">=</span> <span class="st">"CD4_T_cells"</span>,</span>
<span>                                                                    `T_cells_4` <span class="op">=</span> <span class="st">"CD8_T_cells"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; The following `from` values were not present in `x`: T_cells_4</span></span></code></pre></div>
<p>Now that we identified the marker genes describing the different T
cell populations. We can re-annotate them based on their expression
profile and a new prediciton from Celltypist. After this we, can easily
transfer the labels in the subset to the original object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Seu_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.TransferLabel.html">DO.TransferLabel</a></span><span class="op">(</span><span class="va">Seu_obj</span>,</span>
<span>                            Subset_obj <span class="op">=</span> <span class="va">T_cells</span>,</span>
<span>                            annotation_column <span class="op">=</span> <span class="st">"annotation"</span>,</span>
<span>                            subset_annotation <span class="op">=</span> <span class="st">"annotation"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/DO.UMAP.html">DO.UMAP</a></span><span class="op">(</span><span class="va">Seu_obj</span>, group.by <span class="op">=</span> <span class="st">"annotation"</span>, legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/re-annotate-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="gene-ontology-analysis">Gene ontology analysis<a class="anchor" aria-label="anchor" href="#gene-ontology-analysis"></a>
</h3>
<p>To explore which biological processes are enriched in a specific cell
type across conditions, we can perform gene ontology analysis. We’ll
start by identifying differentially expressed genes, focusing here on T
cells. For differential gene expression analysis, we introduced a new
function, which combines DGE analysis using a single cell approach,
e.g. the popular Wilcoxon test and a pseudobulk testing using DESeq2. We
can then observe the results in a combined dataframe.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this data set contains only one sample per condition we introduce replicates for showing the pseudo bulk approach</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">Seu_obj</span><span class="op">$</span><span class="va">orig.ident2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>, <span class="st">"F"</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">Seu_obj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">CD4T_cells</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.Subset.html">DO.Subset</a></span><span class="op">(</span><span class="va">Seu_obj</span>, ident <span class="op">=</span> <span class="st">"annotation"</span>, ident_name <span class="op">=</span> <span class="st">"CD4_T_cells"</span><span class="op">)</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:47 - Specified 'ident_name': expecting a categorical variable.</span></span>
<span></span>
<span><span class="va">DGE_result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.MultiDGE.html">DO.MultiDGE</a></span><span class="op">(</span><span class="va">CD4T_cells</span>,</span>
<span>                          sample_col <span class="op">=</span> <span class="st">"orig.ident2"</span>,</span>
<span>                          method_sc <span class="op">=</span> <span class="st">"wilcox"</span>,</span>
<span>                          ident_ctrl <span class="op">=</span> <span class="st">"healthy"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Names of identity class contain underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This message is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:47 - Corrected annotation names in pseudo-bulk object by replacing '-' with '_'.</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:47 - Starting DGE single cell method analysis</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:47 - Comparing disease with healthy in: CD4_T_cells</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:49 - Finished DGE single cell method analysis</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:49 - Starting DGE pseudo bulk method analysis</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:49 - Comparing disease with healthy in: CD4_T_cells</span></span>
<span><span class="co">#&gt; converting counts to integer mode</span></span>
<span><span class="co">#&gt; gene-wise dispersion estimates</span></span>
<span><span class="co">#&gt; mean-dispersion relationship</span></span>
<span><span class="co">#&gt; final dispersion estimates</span></span>
<span><span class="co">#&gt; 2025-07-26 07:35:50 - Finished DGE pseudo bulk method analysis</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DGE_result</span>,<span class="fl">10</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"style='width:100%;'"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kable_styling</a></span><span class="op">(</span>bootstrap_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"striped"</span>, <span class="st">"hover"</span>, <span class="st">"condensed"</span>, <span class="st">"responsive"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%; margin-left: auto; margin-right: auto;" class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
gene
</th>
<th style="text-align:right;">
pct.1
</th>
<th style="text-align:right;">
pct.2
</th>
<th style="text-align:left;">
celltype
</th>
<th style="text-align:left;">
condition
</th>
<th style="text-align:right;">
avg_log2FC_PB_DESeq2
</th>
<th style="text-align:right;">
avg_log2FC_SC_wilcox
</th>
<th style="text-align:right;">
p_val_adj_PB_DESeq2
</th>
<th style="text-align:right;">
p_val_adj_SC_wilcox
</th>
<th style="text-align:right;">
p_val_PB_DESeq2
</th>
<th style="text-align:right;">
p_val_SC_wilcox
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
RGS1
</td>
<td style="text-align:right;">
0.847
</td>
<td style="text-align:right;">
0.053
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
5.725615
</td>
<td style="text-align:right;">
6.390029
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
SRGN
</td>
<td style="text-align:right;">
0.985
</td>
<td style="text-align:right;">
0.481
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
3.713146
</td>
<td style="text-align:right;">
4.318726
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ZFP36
</td>
<td style="text-align:right;">
0.908
</td>
<td style="text-align:right;">
0.419
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
2.921135
</td>
<td style="text-align:right;">
3.460342
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
KLF6
</td>
<td style="text-align:right;">
0.913
</td>
<td style="text-align:right;">
0.415
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
2.163426
</td>
<td style="text-align:right;">
2.734643
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
FOS
</td>
<td style="text-align:right;">
0.939
</td>
<td style="text-align:right;">
0.583
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
2.338456
</td>
<td style="text-align:right;">
2.944901
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
RGCC
</td>
<td style="text-align:right;">
0.810
</td>
<td style="text-align:right;">
0.326
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
2.686536
</td>
<td style="text-align:right;">
3.148416
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
DUSP2
</td>
<td style="text-align:right;">
0.767
</td>
<td style="text-align:right;">
0.235
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
2.895376
</td>
<td style="text-align:right;">
3.457257
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
DUSP1
</td>
<td style="text-align:right;">
0.937
</td>
<td style="text-align:right;">
0.595
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
1.560897
</td>
<td style="text-align:right;">
2.150082
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
ACTB
</td>
<td style="text-align:right;">
0.985
</td>
<td style="text-align:right;">
0.999
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
-1.948105
</td>
<td style="text-align:right;">
-1.475069
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
<tr>
<td style="text-align:left;">
LMNA
</td>
<td style="text-align:right;">
0.497
</td>
<td style="text-align:right;">
0.049
</td>
<td style="text-align:left;">
CD4_T_cells
</td>
<td style="text-align:left;">
disease
</td>
<td style="text-align:right;">
3.789865
</td>
<td style="text-align:right;">
4.570318
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
</tr>
</tbody>
</table>
<p>After inspecting the DGE analysis, we continue with
<code>DO.enrichR</code> function, which uses the enrichR API to run gene
set enrichment. It separates the DE genes into up- and down-regulated
sets and runs the analysis for each group independently</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result_GO</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DO.enrichR.html">DO.enrichR</a></span><span class="op">(</span>df_DGE <span class="op">=</span> <span class="va">DGE_result</span>,</span>
<span>                        gene_column <span class="op">=</span> <span class="st">"gene"</span>,</span>
<span>                        pval_column <span class="op">=</span> <span class="st">"p_val_adj_SC_wilcox"</span>,</span>
<span>                        log2fc_column <span class="op">=</span> <span class="st">"avg_log2FC_SC_wilcox"</span>,</span>
<span>                        pval_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                        log2fc_cutoff <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                        path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                        filename <span class="op">=</span> <span class="st">""</span>,</span>
<span>                        species <span class="op">=</span> <span class="st">"Human"</span>,</span>
<span>                        go_catgs <span class="op">=</span> <span class="st">"GO_Biological_Process_2023"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Connection changed to https://maayanlab.cloud/Enrichr/</span></span>
<span><span class="co">#&gt; Connection is Live!</span></span>
<span><span class="co">#&gt; Uploading data to Enrichr... Done.</span></span>
<span><span class="co">#&gt;   Querying GO_Biological_Process_2023... Done.</span></span>
<span><span class="co">#&gt; Parsing results... Done.</span></span>
<span><span class="co">#&gt; Uploading data to Enrichr... Done.</span></span>
<span><span class="co">#&gt;   Querying GO_Biological_Process_2023... Done.</span></span>
<span><span class="co">#&gt; Parsing results... Done.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">result_GO</span>,<span class="fl">5</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span>format <span class="op">=</span> <span class="st">"html"</span>, table.attr <span class="op">=</span> <span class="st">"style='width:100%;'"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/kableExtra/man/kable_styling.html" class="external-link">kable_styling</a></span><span class="op">(</span>bootstrap_options <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"striped"</span>, <span class="st">"hover"</span>, <span class="st">"condensed"</span>, <span class="st">"responsive"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%; margin-left: auto; margin-right: auto;" class="table table table-striped table-hover table-condensed table-responsive">
<thead><tr>
<th style="text-align:left;">
Term
</th>
<th style="text-align:left;">
Overlap
</th>
<th style="text-align:right;">
P.value
</th>
<th style="text-align:right;">
Adjusted.P.value
</th>
<th style="text-align:right;">
Old.P.value
</th>
<th style="text-align:right;">
Old.Adjusted.P.value
</th>
<th style="text-align:right;">
Odds.Ratio
</th>
<th style="text-align:right;">
Combined.Score
</th>
<th style="text-align:left;">
Genes
</th>
<th style="text-align:left;">
Database
</th>
<th style="text-align:left;">
State
</th>
</tr></thead>
<tbody>
<tr>
<td style="text-align:left;">
Positive Regulation Of Cytokine Production (<a href="GO:0001819" class="uri">GO:0001819</a>)
</td>
<td style="text-align:left;">
23/320
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0e+00
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
9.103525
</td>
<td style="text-align:right;">
279.79161
</td>
<td style="text-align:left;">
EGR1;CD74;CEBPB;ANXA1;F2R;TWIST1;RORA;ISG15;TNF;SLC7A5;RGCC;PTPRC;IFNG;IRF4;NLRP3;CCR7;IL6ST;TIGIT;GAPDH;CD200;HSPA1B;HLA-DPA1;HSPA1A
</td>
<td style="text-align:left;">
GO_Biological_Process_2023
</td>
<td style="text-align:left;">
enriched
</td>
</tr>
<tr>
<td style="text-align:left;">
Negative Regulation Of Transcription By RNA Polymerase II (<a href="GO:0000122" class="uri">GO:0000122</a>)
</td>
<td style="text-align:left;">
29/763
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
2e-07
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4.710771
</td>
<td style="text-align:right;">
106.17869
</td>
<td style="text-align:left;">
CEBPA;CEBPB;CITED2;HMGB2;TWIST1;PRDM1;TNF;ZFP36;TRPS1;KLF10;JUN;CAV1;IRF2BP2;KLF4;SAP30;NR4A2;ETV7;ZEB2;MAF;NR4A3;BCL6;IFNG;ID2;ID1;BHLHE40;ID3;FOSB;ATF3;HSPA1A
</td>
<td style="text-align:left;">
GO_Biological_Process_2023
</td>
<td style="text-align:left;">
enriched
</td>
</tr>
<tr>
<td style="text-align:left;">
Regulation Of Transcription By RNA Polymerase II (<a href="GO:0006357" class="uri">GO:0006357</a>)
</td>
<td style="text-align:left;">
49/2028
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3e-07
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3.153714
</td>
<td style="text-align:right;">
67.92806
</td>
<td style="text-align:left;">
CITED2;HMGB2;RORA;PRDM1;IKZF2;TNF;ZFP36;TRPS1;NAMPT;KDM6B;KLF10;TET2;FOS;ETV3;SAP30;ETV7;TOX2;ZEB2;MAF;IFNG;IRF4;ATF3;CEBPA;CEBPB;TWIST1;RBBP8;NLRP3;HES4;EGR1;JUN;EGR3;CAV1;IRF2BP2;KLF4;FOSL2;NR4A2;NFKBIA;NR4A1;KLF6;RGCC;NR4A3;BCL6;ID2;ID1;BHLHE40;REL;ID3;FOSB;HSPA1A
</td>
<td style="text-align:left;">
GO_Biological_Process_2023
</td>
<td style="text-align:left;">
enriched
</td>
</tr>
<tr>
<td style="text-align:left;">
Regulation Of DNA-templated Transcription (<a href="GO:0006355" class="uri">GO:0006355</a>)
</td>
<td style="text-align:left;">
47/1922
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
4e-07
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3.166167
</td>
<td style="text-align:right;">
66.49666
</td>
<td style="text-align:left;">
CDKN1C;CEBPA;CEBPB;CITED2;HMGB2;TWIST1;RORA;PRDM1;IKZF2;TNF;ZFP36L2;BASP1;TRPS1;RBBP8;TNFRSF4;HES4;KLF10;EGR1;CD74;JUN;EGR3;F2R;IRF2BP2;FOS;ETV3;KLF4;SAP30;FOSL2;NR4A2;ETV7;NR4A1;TOX2;ZEB2;KLF6;MAF;NR4A3;BCL6;IFNG;IRF4;ID2;ID1;BHLHE40;REL;ID3;FOSB;ATF3;HLA-DRB1
</td>
<td style="text-align:left;">
GO_Biological_Process_2023
</td>
<td style="text-align:left;">
enriched
</td>
</tr>
<tr>
<td style="text-align:left;">
Negative Regulation Of DNA-templated Transcription (<a href="GO:0045892" class="uri">GO:0045892</a>)
</td>
<td style="text-align:left;">
32/1025
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
9e-07
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
0
</td>
<td style="text-align:right;">
3.862554
</td>
<td style="text-align:right;">
76.86009
</td>
<td style="text-align:left;">
CDKN1C;CEBPA;CEBPB;CITED2;HMGB2;TWIST1;PRDM1;TNF;ZFP36;BASP1;TRPS1;TNFRSF4;KLF10;JUN;CAV1;IRF2BP2;KLF4;SAP30;NR4A2;ETV7;ZEB2;MAF;NR4A3;BCL6;IFNG;ID2;ID1;BHLHE40;ID3;FOSB;ATF3;HSPA1A
</td>
<td style="text-align:left;">
GO_Biological_Process_2023
</td>
<td style="text-align:left;">
enriched
</td>
</tr>
</tbody>
</table>
<p>The top significant results can then be visualized in a bar plot.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">result_GO_sig</span> <span class="op">&lt;-</span> <span class="va">result_GO</span><span class="op">[</span><span class="va">result_GO</span><span class="op">$</span><span class="va">Adjusted.P.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span>, <span class="op">]</span> </span>
<span><span class="va">result_GO_sig</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="st">"CD4T_cells"</span></span>
<span></span>
<span><span class="fu"><a href="../reference/DO.SplitBarGSEA.html">DO.SplitBarGSEA</a></span><span class="op">(</span>df_GSEA <span class="op">=</span> <span class="va">result_GO_sig</span>,</span>
<span>                term_col <span class="op">=</span> <span class="st">"Term"</span>,</span>
<span>                col_split <span class="op">=</span> <span class="st">"Combined.Score"</span>,</span>
<span>                cond_col <span class="op">=</span> <span class="st">"State"</span>,</span>
<span>                pos_cond <span class="op">=</span> <span class="st">"enriched"</span>,</span>
<span>                showP <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">base</span>, <span class="st">"/"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">GSEA_plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path <span class="op">=</span> <span class="va">base</span>, pattern <span class="op">=</span> <span class="st">"SplitBar.*\\.svg$"</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">magick</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/magick/reference/editing.html" class="external-link">image_read_svg</a></span><span class="op">(</span><span class="va">GSEA_plot</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="DOtools_files/figure-html/GO%20visualisation-1.png" width="960"></p>
</div>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  C.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    C.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2025-07-26</span></span>
<span><span class="co">#&gt;  pandoc   3.1.11 @ /opt/hostedtoolcache/pandoc/3.1.11/x64/ (via rmarkdown)</span></span>
<span><span class="co">#&gt;  quarto   NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package             </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  abind                  1.4-8    <span style="color: #555555;">2024-09-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  backports              1.5.0    <span style="color: #555555;">2024-05-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  basilisk               1.20.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  basilisk.utils         1.20.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  beachmat               2.24.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Biobase                2.68.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocGenerics           0.54.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocManager            1.30.26  <span style="color: #555555;">2025-06-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  BiocParallel           1.42.1   <span style="color: #555555;">2025-06-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  BiocStyle            * 2.36.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  bookdown               0.43     <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  broom                  1.0.8    <span style="color: #555555;">2025-03-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  bslib                  0.9.0    <span style="color: #555555;">2025-01-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cachem                 1.1.0    <span style="color: #555555;">2024-05-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  car                    3.1-3    <span style="color: #555555;">2024-09-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  carData                3.0-5    <span style="color: #555555;">2022-01-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cli                    3.6.5    <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cluster                2.1.8.1  <span style="color: #555555;">2025-03-12</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  codetools              0.2-20   <span style="color: #555555;">2024-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  colorspace             2.1-1    <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  cowplot                1.2.0    <span style="color: #555555;">2025-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  crayon                 1.5.3    <span style="color: #555555;">2024-06-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  curl                   6.4.0    <span style="color: #555555;">2025-06-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  data.table             1.17.8   <span style="color: #555555;">2025-07-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DelayedArray           0.34.1   <span style="color: #555555;">2025-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  DelayedMatrixStats     1.30.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  deldir                 2.0-4    <span style="color: #555555;">2024-02-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  desc                   1.4.3    <span style="color: #555555;">2023-12-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DESeq2                 1.48.1   <span style="color: #555555;">2025-05-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  digest                 0.6.37   <span style="color: #555555;">2024-08-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dir.expiry             1.16.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  dotCall64              1.2      <span style="color: #555555;">2024-10-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DOtools              * 0.99.0   <span style="color: #555555;">2025-07-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">local</span></span></span>
<span><span class="co">#&gt;  dplyr                * 1.1.4    <span style="color: #555555;">2023-11-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  dqrng                  0.4.1    <span style="color: #555555;">2024-05-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  DropletUtils           1.28.1   <span style="color: #555555;">2025-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  edgeR                  4.6.3    <span style="color: #555555;">2025-07-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  enrichR              * 3.4      <span style="color: #555555;">2025-02-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  evaluate               1.0.4    <span style="color: #555555;">2025-06-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  farver                 2.1.2    <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastDummies            1.7.5    <span style="color: #555555;">2025-01-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fastmap                1.2.0    <span style="color: #555555;">2024-05-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  filelock               1.0.3    <span style="color: #555555;">2023-12-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fitdistrplus           1.2-4    <span style="color: #555555;">2025-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Formula                1.2-5    <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  fs                     1.6.6    <span style="color: #555555;">2025-04-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  future               * 1.58.0   <span style="color: #555555;">2025-06-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  future.apply           1.20.0   <span style="color: #555555;">2025-06-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  generics               0.1.4    <span style="color: #555555;">2025-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb           1.44.1   <span style="color: #555555;">2025-07-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData       1.2.14   <span style="color: #555555;">2025-06-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomicRanges          1.60.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  ggalluvial             0.12.5   <span style="color: #555555;">2023-02-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggcorrplot             0.1.4.1  <span style="color: #555555;">2023-09-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggiraph                0.8.13   <span style="color: #555555;">2025-03-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggiraphExtra           0.3.0    <span style="color: #555555;">2020-10-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggplot2                3.5.2    <span style="color: #555555;">2025-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggpubr                 0.6.1    <span style="color: #555555;">2025-06-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggrepel                0.9.6    <span style="color: #555555;">2024-09-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggridges               0.5.6    <span style="color: #555555;">2024-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggsignif               0.6.4    <span style="color: #555555;">2022-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ggtext                 0.1.2    <span style="color: #555555;">2022-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  globals                0.18.0   <span style="color: #555555;">2025-05-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  glue                   1.8.0    <span style="color: #555555;">2024-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  goftest                1.2-3    <span style="color: #555555;">2021-10-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gridExtra              2.3      <span style="color: #555555;">2017-09-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gridtext               0.1.5    <span style="color: #555555;">2022-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  gtable                 0.3.6    <span style="color: #555555;">2024-10-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  h5mread                1.0.1    <span style="color: #555555;">2025-05-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  HDF5Array              1.36.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  hms                    1.1.3    <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmltools              0.5.8.1  <span style="color: #555555;">2024-04-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  htmlwidgets            1.6.4    <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httpuv                 1.6.16   <span style="color: #555555;">2025-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  httr                   1.4.7    <span style="color: #555555;">2023-08-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ica                    1.0-3    <span style="color: #555555;">2022-07-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  igraph                 2.1.4    <span style="color: #555555;">2025-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  insight                1.3.1    <span style="color: #555555;">2025-06-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  IRanges                2.42.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  irlba                  2.3.5.1  <span style="color: #555555;">2022-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jquerylib              0.1.4    <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  jsonlite               2.0.0    <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  kableExtra           * 1.4.0    <span style="color: #555555;">2024-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  KernSmooth             2.23-26  <span style="color: #555555;">2025-01-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  knitr                  1.50     <span style="color: #555555;">2025-03-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  labeling               0.4.3    <span style="color: #555555;">2023-08-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  later                  1.4.2    <span style="color: #555555;">2025-04-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lattice                0.22-7   <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  lazyeval               0.2.2    <span style="color: #555555;">2019-03-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  leidenbase             0.1.35   <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lifecycle              1.0.4    <span style="color: #555555;">2023-11-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  limma                  3.64.1   <span style="color: #555555;">2025-05-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  listenv                0.9.1    <span style="color: #555555;">2024-01-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  lmtest                 0.9-40   <span style="color: #555555;">2022-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  locfit                 1.5-9.12 <span style="color: #555555;">2025-03-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magick                 2.8.7    <span style="color: #555555;">2025-06-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  magrittr               2.0.3    <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  MASS                   7.3-65   <span style="color: #555555;">2025-02-28</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  Matrix                 1.7-3    <span style="color: #555555;">2025-03-11</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics         1.20.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  matrixStats            1.5.0    <span style="color: #555555;">2025-01-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mgcv                   1.9-3    <span style="color: #555555;">2025-04-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  mime                   0.13     <span style="color: #555555;">2025-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  miniUI                 0.1.2    <span style="color: #555555;">2025-04-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  mycor                  0.1.1    <span style="color: #555555;">2018-04-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  nlme                   3.1-168  <span style="color: #555555;">2025-03-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  openxlsx               4.2.8    <span style="color: #555555;">2025-01-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  parallelly             1.45.1   <span style="color: #555555;">2025-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  patchwork              1.3.1    <span style="color: #555555;">2025-06-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pbapply                1.7-4    <span style="color: #555555;">2025-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pillar                 1.11.0   <span style="color: #555555;">2025-07-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgconfig              2.0.3    <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  pkgdown                2.1.3    <span style="color: #555555;">2025-05-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">any (@2.1.3)</span></span></span>
<span><span class="co">#&gt;  plotly                 4.11.0   <span style="color: #555555;">2025-06-19</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  plyr                 * 1.8.9    <span style="color: #555555;">2023-10-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  png                    0.1-8    <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  polyclip               1.10-7   <span style="color: #555555;">2024-07-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ppcor                  1.1      <span style="color: #555555;">2015-12-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  prettyunits            1.2.0    <span style="color: #555555;">2023-09-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  progress               1.2.3    <span style="color: #555555;">2023-12-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  progressr              0.15.1   <span style="color: #555555;">2024-11-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  promises               1.3.3    <span style="color: #555555;">2025-05-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  purrr                  1.1.0    <span style="color: #555555;">2025-07-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.methodsS3            1.8.2    <span style="color: #555555;">2022-06-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.oo                   1.27.1   <span style="color: #555555;">2025-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R.utils                2.13.0   <span style="color: #555555;">2025-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  R6                     2.6.1    <span style="color: #555555;">2025-02-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ragg                   1.4.0    <span style="color: #555555;">2025-04-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RANN                   2.6.2    <span style="color: #555555;">2024-08-25</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RColorBrewer           1.1-3    <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rcpp                   1.1.0    <span style="color: #555555;">2025-07-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RcppAnnoy              0.0.22   <span style="color: #555555;">2024-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RcppHNSW               0.6.0    <span style="color: #555555;">2024-02-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reshape2               1.4.4    <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  reticulate             1.43.0   <span style="color: #555555;">2025-07-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rhdf5                  2.52.1   <span style="color: #555555;">2025-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  rhdf5filters           1.20.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  Rhdf5lib               1.30.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  rjson                  0.2.23   <span style="color: #555555;">2024-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rlang                  1.1.6    <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rmarkdown              2.29     <span style="color: #555555;">2024-11-04</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  ROCR                   1.0-11   <span style="color: #555555;">2020-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  RSpectra               0.16-2   <span style="color: #555555;">2024-07-18</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rstatix                0.7.2    <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rstudioapi             0.17.1   <span style="color: #555555;">2024-10-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  rsvg                   2.6.2    <span style="color: #555555;">2025-03-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  Rtsne                  0.17     <span style="color: #555555;">2023-12-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  S4Arrays               1.8.1    <span style="color: #555555;">2025-06-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  S4Vectors              0.46.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sass                   0.4.10   <span style="color: #555555;">2025-04-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scales                 1.4.0    <span style="color: #555555;">2025-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scattermore            1.2      <span style="color: #555555;">2023-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sctransform            0.4.2    <span style="color: #555555;">2025-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  scuttle                1.18.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sessioninfo            1.2.3    <span style="color: #555555;">2025-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">any (@1.2.3)</span></span></span>
<span><span class="co">#&gt;  Seurat               * 5.3.0    <span style="color: #555555;">2025-04-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  SeuratObject         * 5.1.0    <span style="color: #555555;">2025-04-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  shiny                  1.11.1   <span style="color: #555555;">2025-07-03</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  SingleCellExperiment   1.30.1   <span style="color: #555555;">2025-05-07</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sjlabelled             1.2.0    <span style="color: #555555;">2022-04-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sjmisc                 2.8.10   <span style="color: #555555;">2024-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  sp                   * 2.2-0    <span style="color: #555555;">2025-02-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spam                   2.11-1   <span style="color: #555555;">2025-01-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  SparseArray            1.8.1    <span style="color: #555555;">2025-07-23</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  sparseMatrixStats      1.20.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  spatstat.data          3.1-6    <span style="color: #555555;">2025-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.explore       3.5-2    <span style="color: #555555;">2025-07-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.geom          3.5-0    <span style="color: #555555;">2025-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.random        3.4-1    <span style="color: #555555;">2025-05-20</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.sparse        3.1-0    <span style="color: #555555;">2024-06-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.univar        3.1-4    <span style="color: #555555;">2025-07-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  spatstat.utils         3.1-5    <span style="color: #555555;">2025-07-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  statmod                1.5.0    <span style="color: #555555;">2023-01-06</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringi                1.8.7    <span style="color: #555555;">2025-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  stringr                1.5.1    <span style="color: #555555;">2023-11-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment   1.38.1   <span style="color: #555555;">2025-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  survival               3.8-3    <span style="color: #555555;">2024-12-17</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.5.1)</span></span></span>
<span><span class="co">#&gt;  svglite                2.2.1    <span style="color: #555555;">2025-05-12</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  systemfonts            1.2.3    <span style="color: #555555;">2025-04-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tensor                 1.5.1    <span style="color: #555555;">2025-06-17</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  textshaping            1.0.1    <span style="color: #555555;">2025-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tibble               * 3.3.0    <span style="color: #555555;">2025-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyr                  1.3.1    <span style="color: #555555;">2024-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyselect             1.2.1    <span style="color: #555555;">2024-03-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  tidyverse              2.0.0    <span style="color: #555555;">2023-02-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  UCSC.utils             1.4.0    <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  uuid                   1.2-1    <span style="color: #555555;">2024-07-29</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  uwot                   0.2.3    <span style="color: #555555;">2025-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  vctrs                  0.6.5    <span style="color: #555555;">2023-12-01</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  viridisLite            0.4.2    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  withr                  3.0.2    <span style="color: #555555;">2024-10-28</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  WriteXLS               6.8.0    <span style="color: #555555;">2025-05-22</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xfun                   0.52     <span style="color: #555555;">2025-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xml2                   1.3.8    <span style="color: #555555;">2025-03-14</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  xtable                 1.8-4    <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  XVector                0.48.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  yaml                   2.3.10   <span style="color: #555555;">2024-07-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zellkonverter          1.18.0   <span style="color: #555555;">2025-04-15</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Bioconduc~</span></span></span>
<span><span class="co">#&gt;  zip                    2.3.3    <span style="color: #555555;">2025-05-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt;  zoo                    1.8-14   <span style="color: #555555;">2025-04-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /home/runner/work/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /opt/R/4.5.1/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /opt/R/4.5.1/lib/R/library</span></span></span>
<span><span class="co">#&gt;  <span style="color: #BBBBBB; background-color: #BB0000;">*</span> ── Packages attached to the search path.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Python configuration ───────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  python:         /home/runner/.cache/R/basilisk/1.20.0/zellkonverter/1.18.0/zellkonverterAnnDataEnv-0.11.4/bin/python</span></span>
<span><span class="co">#&gt;  libpython:      /home/runner/.cache/R/basilisk/1.20.0/zellkonverter/1.18.0/zellkonverterAnnDataEnv-0.11.4/lib/libpython3.13.so</span></span>
<span><span class="co">#&gt;  pythonhome:     /home/runner/.cache/R/basilisk/1.20.0/zellkonverter/1.18.0/zellkonverterAnnDataEnv-0.11.4:/home/runner/.cache/R/basilisk/1.20.0/zellkonverter/1.18.0/zellkonverterAnnDataEnv-0.11.4</span></span>
<span><span class="co">#&gt;  version:        3.13.2 | packaged by conda-forge | (main, Feb 17 2025, 14:10:22) [GCC 13.3.0]</span></span>
<span><span class="co">#&gt;  numpy:          /home/runner/.cache/R/basilisk/1.20.0/zellkonverter/1.18.0/zellkonverterAnnDataEnv-0.11.4/lib/python3.13/site-packages/numpy</span></span>
<span><span class="co">#&gt;  numpy_version:  2.2.4</span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;  NOTE: Python version was forced by use_python() function</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Mariano Ruz Jurado, David Rodriguez Morales, David John.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
